src/ATen/native/xpu/sycl/RNNKernels.cpp

```diff
@@ -0,0 +1,433 @@
+      grad_hx_[index_] = grad_hy_[index_] * workspace_[workspacez_index + loc];
+    }
+  }
+  FuseOpsKernelBackwardFunctor(
+      const T* grad_hy,
+      const T* workspace,
+      T* grad_input_gates,
+      T* grad_hidden_gates,
+      T* grad_hx,
+      const int64_t feature_size,
+      const int64_t batch_size,
+      const int64_t COL,
+      const int64_t ROW,
+      int64_t GROUP_SIZE)
+      : grad_hy_(grad_hy),
+        workspace_(workspace),
+        grad_input_gates_(grad_input_gates),
+        grad_hidden_gates_(grad_hidden_gates),
+        grad_hx_(grad_hx),
+        feature_size_(feature_size),
+        batch_size_(batch_size),
+        COL_(COL),
+        ROW_(ROW),
+        GROUP_SIZE_(GROUP_SIZE) {}
+
+ private:
+  const T* grad_hy_;
+  const T* workspace_;
+  T* grad_input_gates_;
+  T* grad_hidden_gates_;
+  T* grad_hx_;
+  const int64_t feature_size_;
+  const int64_t batch_size_;
+  const int64_t COL_;
+  const int64_t ROW_;
+  int64_t GROUP_SIZE_;
+};
+
+// backward sycl implementation
+template <typename T>
+static inline void fuse_ops_kernel_backward(
+    const T* grad_hy,
+    const T* workspace,
+    T* grad_input_gates,
+    T* grad_hidden_gates,
+    T* grad_hx,
+    const int64_t feature_size,
+    const int64_t batch_size) {
+  using KernelClass = FuseOpsKernelBackwardFunctor<T>;
+  int64_t max_group_size = syclMaxWorkGroupSize<KernelClass>();
+  const int64_t COL = feature_size;
+  const int64_t ROW = batch_size;
+  int64_t group_size = std::min(feature_size, max_group_size);
+
+  const sycl::range<1> global_size{(size_t)ROW * group_size};
+  const sycl::range<1> local_size{(size_t)group_size};
+  // all input gates have the same shape
+  // all intermdiate variables are precessed in this function
+
+  KernelClass kfn(
+      grad_hy,
+      workspace,
+      grad_input_gates,
+      grad_hidden_gates,
+      grad_hx,
+      feature_size,
+      batch_size,
+      COL,
+      ROW,
+      group_size);
+
+  sycl_kernel_submit(global_size, local_size, getCurrentSYCLQueue(), kfn);
+}
+
+std::tuple<Tensor, Tensor> _thnn_fused_gru_cell_kernel(
+    const Tensor& input_gates,
+    const Tensor& hidden_gates,
+    const Tensor& hx,
+    const c10::optional<Tensor>& input_bias_opt,
+    const c10::optional<Tensor>& hidden_bias_opt) {
+  c10::MaybeOwned<Tensor> input_bias_maybe_owned =
+      at::borrow_from_optional_tensor(input_bias_opt);
+  const Tensor& input_bias = *input_bias_maybe_owned;
+  const Tensor& hidden_bias =
+      c10::value_or_else(hidden_bias_opt, [] { return Tensor(); });
+
+  auto batched_input = true;
+  auto feature_size = hx.size(1);
+  auto batch_size = hx.size(0);
+
+  if (hx.dim() == 1) {
+    // no batch input
+    hx.resize_({1, hx.size(0)});
+  }
+
+  at::Tensor workspace = at::empty(
+      {hx.size(0), hx.size(1) * GRU_WORKSPACE_MULTIPLIER}, hx.options());
+
+  auto hy = at::empty_like(hx, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+
+  AT_DISPATCH_FLOATING_TYPES_AND2(
```

Comment:
How we handle non-contiguous input ? CUDA uses tensorInfo, but we just use raw data pointer, without stride information